### PR TITLE
#2269 Add validation rules for wrong image path

### PR DESCRIPTION
**Pull request by lfasani**
Two new validation rules are added:
* I_46 - Check that the image used for diagram nodes is found
* I_47 - Check that images used in rich text is found (Capella and
diagram descriptions etc)

The corresponding quick fix is also added.

Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  protected Logger logger = ReportManagerRegistry.getInstance().subscribe(IReportManagerDefaultComponents.VALIDATION);
+
+  protected StringBuilder desc = null;
+
+  @Override
+  public IStatus validate(final IValidationContext ctx) {
+    // Ugly code to check only if called by the Capella Validation
+    // Indeed, otherwise, this code is called many times by Sirius and it is too impacting to change Sirius.
+    boolean calledByCapella = Arrays.asList(Thread.getAllStackTraces().get(Thread.currentThread())).stream()
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java:53]: Can you replace `Arrays.asList(Thread.getAllStackTraces().get(Thread.currentThread())).stream()` by 
`Stream.of(Thread.currentThread().getStackTrace())`

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * Check if the images used on diagram elements, can be found.</br>
+ * Warning: This rule will load all representation associated to the validation context(Capella element)
+ * 
+ * @author lfasani
+ */
+public class ImagePathInDiagramCheck extends ImagePathConstraint {
+
+  protected Logger logger = ReportManagerRegistry.getInstance().subscribe(IReportManagerDefaultComponents.VALIDATION);
+
+  protected StringBuilder desc = null;
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java:47]: Can you remove both unused desc and logger

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * Check if images can be found for images used in the rich text description of the capella element, diagram description
+ * or other rich text attributes.
+ * 
+ * @author lfasani
+ */
+@SuppressWarnings("restriction")
+public class ImagePathOnRichTextAttributeCheck extends ImagePathConstraint {
+
+  protected StringBuilder desc = null;
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java:41]: same

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Override
+  public IStatus validate(final IValidationContext ctx) {
+    // Ugly code to check only if called by the Capella Validation
+    // Indeed, otherwise, this code is called many times by Sirius and it is too impacting to change Sirius.
+    boolean calledByCapella = Arrays.asList(Thread.getAllStackTraces().get(Thread.currentThread())).stream()
+        .filter(ste -> ste.getClassName().equals(CapellaDiagnostician.class.getName())).findFirst().isPresent();
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java:48]: same

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathResolver.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      }
+    }
+  }
+
+  /**
+   * Browse the model upward (from the leaf to the root) and return the first representation found.
+   * 
+   * @return the representation if found, null otherwise.
+   */
+  private DRepresentation getRepresentation(EObject eObject) {
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathResolver.java:112]: can you replace EObject by DSemanticDecorator

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/plugin.xml [hunk 6]
```diff
@@ -98,6 +114,44 @@
...
+                  Warning: This rule applies on diagrams elements of diagram associated to the Capella element and will load the diagram. 
+It is recommended to uncheck this rule by default.
+
+This rule ensures that images, used as background of diagram elements, can be found.
+               </description>
+            </constraint>
+            <constraint
+                  class="org.polarsys.capella.core.platform.sirius.sirius.validation.ddiagram.ImagePathOnRichTextAttributeCheck"
+                  id="I_47"
+                  isEnabledByDefault="false"
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/plugin.xml:142]: Can you enable this one by default.

**Review by pdulth — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java (hunk 1 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java:53]: done

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java (hunk 2 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathInDiagramCheck.java:47]: Done

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java (hunk 4 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathOnRichTextAttributeCheck.java:48]: Done

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathResolver.java (hunk 5 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/src/org/polarsys/capella/core/platform/sirius/sirius/validation/ddiagram/ImagePathResolver.java:112]: Done

File: core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/plugin.xml (hunk 6 above)
Comment (lfasani) [reply core/plugins/org.polarsys.capella.core.platform.sirius.sirius.validation/plugin.xml:142]: Done

**Review by pdulth — APPROVED** [review]

**Comment by pdulth** [comment]
Can you look at failing related junit. `ImagePathInRichTextAttributeMigrationTest.ImagePathInRichTextAttributeMigrationTest`

**Comment by lfasani** [comment]
> Can you look at failing related junit. `ImagePathInRichTextAttributeMigrationTest.ImagePathInRichTextAttributeMigrationTest`

Yes I am on it. I'll do another commit because I also need to improve the log during the migration

**Pull request merged**